reference/matrix/csr_kernels.cpp

```diff
@@ -1118,6 +1118,160 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
 
 
+template <typename IndexType>
+void build_lookup_offsets(std::shared_ptr<const ReferenceExecutor> exec,
+                          const IndexType* row_ptrs, const IndexType* col_idxs,
+                          size_type num_rows,
+                          matrix::csr::sparsity_type allowed,
+                          IndexType* storage_offsets)
+{
+    using matrix::csr::sparsity_bitmap_block_size;
+    using matrix::csr::sparsity_type;
+    for (size_type row = 0; row < num_rows; row++) {
+        const auto row_begin = row_ptrs[row];
+        const auto row_len = row_ptrs[row + 1] - row_begin;
+        const auto local_cols = col_idxs + row_begin;
+        const auto min_col = row_len > 0 ? local_cols[0] : 0;
+        const auto col_range =
+            row_len > 0 ? local_cols[row_len - 1] - min_col + 1 : 0;
+        if (csr_lookup_allowed(allowed, sparsity_type::full) &&
+            row_len == col_range) {
+            storage_offsets[row] = 0;
+        } else {
+            const auto hashmap_storage = std::max<IndexType>(2 * row_len, 1);
+            const auto bitmap_num_blocks = static_cast<int32>(
+                ceildiv(col_range, sparsity_bitmap_block_size));
+            const auto bitmap_storage = 2 * bitmap_num_blocks;
+            if (csr_lookup_allowed(allowed, sparsity_type::bitmap) &&
+                bitmap_storage <= hashmap_storage) {
+                storage_offsets[row] = bitmap_storage;
+            } else {
+                storage_offsets[row] = hashmap_storage;
+            }
+        }
+    }
+    components::prefix_sum(exec, storage_offsets, num_rows + 1);
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(
+    GKO_DECLARE_CSR_BUILD_LOOKUP_OFFSETS_KERNEL);
+
+
+template <typename IndexType>
+bool csr_lookup_try_full(IndexType row_len, IndexType col_range,
+                         matrix::csr::sparsity_type allowed, int64& row_desc)
+{
+    using matrix::csr::sparsity_type;
+    bool is_allowed = csr_lookup_allowed(allowed, sparsity_type::full);
+    if (is_allowed && row_len == col_range) {
+        row_desc = static_cast<int64>(sparsity_type::full);
+        return true;
+    }
+    return false;
+}
+
+
+template <typename IndexType>
+bool csr_lookup_try_bitmap(IndexType row_len, IndexType col_range,
+                           IndexType min_col, IndexType available_storage,
+                           matrix::csr::sparsity_type allowed, int64& row_desc,
+                           int32* local_storage, const IndexType* cols)
+{
+    using matrix::csr::sparsity_bitmap_block_size;
+    using matrix::csr::sparsity_type;
+    bool is_allowed = csr_lookup_allowed(allowed, sparsity_type::bitmap);
+    const auto num_blocks =
+        static_cast<int32>(ceildiv(col_range, sparsity_bitmap_block_size));
+    if (is_allowed && num_blocks * 2 <= available_storage) {
+        row_desc = (static_cast<int64>(num_blocks) << 32) |
+                   static_cast<int64>(sparsity_type::bitmap);
```

Comment:
This cast is a bit suspect because if `sparsity_type` is indeed 64bit, you overwrite your first 32bit. Additionally, this is the only place where you cast it to a `int64`, in all other kernels, you cast it to `int`.